### DISCUSSION TITLE
Bit-level support?

### DISCUSSION THREAD

**Discussion by DerZade** [jam1garner/binrw#222] [Q&A]
I am looking for a library to read / write data with many fields that are only a few bits (not bytes) long. For example, I have 12-bit wide integer fields and therefore many of the fields are not even byte-aligned. 

As far as I can tell that is not really possible with `binrw` atm. Am I correct in this assumption?

Will that be something that will be supported in the (near) future?

**Comment by scarletcafe** [comment]
The [modular-bitfield](https://docs.rs/modular-bitfield) crate sounds like it might satisfy what you need here, and it can be integrated with binrw, there's [an example](https://docs.rs/binrw/0.11.1/binrw/docs/attribute/index.html#using-map-on-a-struct-to-create-a-bit-field) of this in the documentation.

**Comment by DerZade** [comment]
Thought that only really works with byte aligned bit-fields 🤔 (see [here](https://github.com/jam1garner/binrw/discussions/184#discussioncomment-5282443)), but I will try 🙈 

_PS: The issue with Deku for me is that it does not support `Read` / `Write`._

**Comment by csnover** [comment]
This is an old issue but figured I would mention some thoughts for anyone else in future.

bitstreams really are not the best fit for binrw but you can probably make it work with a stream wrapper.

If the objects are always fixed-size but are just not byte-aligned, you could write a generic stream wrapper that tracks the position and injects extra bits at the end of each object to fake alignment to a byte boundary.

If the objects are not fixed-size, you could signal when an object has finished parsing with a marker field that sends some signal back to the stream wrapper to inject the extra bits.

Neither of these are likely to be particularly fast but maybe that is OK for your use case. Otherwise you would need to write a stream buffer that understands the context of what is currently being parsed, at which point you are just hand-writing a parser anyway and binrw isn’t useful.

If someone does decide to make a generic stream solution for this, submitting it as a PR would be cool.
